File: typescript/leet-code/max_depth_of_binary_tree.ts
/**
 * Given the root of a binary tree, return its maximum depth.
 * A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
 *
 * Example 1:
 *                      3
 *                    /   \
 *                   9     20
 *                        /  \
 *                      15     7
 * Input: root = [3,9,20,null,null,15,7]
 * Output: 3
 *
 * Example 2:
 *                      1
 *                       \
 *                         2
 * Input: root = [1,null,2]
 * Output: 2
 *
 * Constraints:
 * The number of nodes in the tree is in the range [0, 104].
 * -100 <= Node.val <= 100
 *
 */

/**
 * Definition for a binary tree node.*/
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function maxDepth(root: TreeNode | null): number {
    if (root === null) {
        return 0;
    }

    let maxChildBranch = Math.max(maxDepth(root.left), maxDepth(root.right));

    return 1 + maxChildBranch;
}

function runDemo() {
    let treeA = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
    let treeB = new TreeNode(1, null, new TreeNode(2));

    console.log(maxDepth(treeA));
    console.log(maxDepth(treeB));
}

runDemo();

export {};

File: rust/rust-coding-challenges/src/leetcode/max_depth_of_binary_tree.rs
use std::cell::RefCell;
use std::rc::Rc;
use std::cmp;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    match root {
        None => 0,
        Some(node) => {
            return  1 + cmp::max(max_depth(node.borrow().left.clone()), max_depth(node.borrow().right.clone()));
        }
    }
}

pub fn run_demo() {
    let tree_a = Some(Rc::new(RefCell::new(TreeNode{
        val: 3,
        left: Some(Rc::new(RefCell::new(TreeNode::new(9)))),
        right: Some(Rc::new(RefCell::new(TreeNode{
            val: 20,
            left: Some(Rc::new(RefCell::new(TreeNode::new(15)))),
            right: Some(Rc::new(RefCell::new(TreeNode::new(7))))
        })))
    })));

    let tree_b = Some(Rc::new(RefCell::new(TreeNode {
        val: 1,
        left: None,
        right: Some(Rc::new(RefCell::new(TreeNode::new(2))))
    })));

    println!("{}", max_depth(tree_a));
    println!("{}", max_depth(tree_b));
}
